# Segmenting and Clustering Neighborhoods in Toronto 

For this assignment, you will be required to explore and cluster the neighborhoods in Toronto.

Start by creating a new Notebook for this assignment.
Use the Notebook to build the code to scrape the following Wikipedia page, https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M, in order to obtain the data that is in the table of postal codes and to transform the data into a pandas dataframe like the one shown below:

3. To create the above dataframe:

The dataframe will consist of three columns: PostalCode, Borough, and Neighborhood
Only process the cells that have an assigned borough. Ignore cells with a borough that is Not assigned.
More than one neighborhood can exist in one postal code area. For example, in the table on the Wikipedia page, you will notice that M5A is listed twice and has two neighborhoods: Harbourfront and Regent Park. These two rows will be combined into one row with the neighborhoods separated with a comma as shown in row 11 in the above table.
If a cell has a borough but a Not assigned neighborhood, then the neighborhood will be the same as the borough. So for the 9th cell in the table on the Wikipedia page, the value of the Borough and the Neighborhood columns will be Queen's Park.
Clean your Notebook and add Markdown cells to explain your work and any assumptions you are making.
In the last cell of your notebook, use the .shape method to print the number of rows of your dataframe

### Import Libraries

In [1]:
## study notes: https://towardsdatascience.com/web-scraping-html-tables-with-python-c9baba21059
import requests
import lxml.html as lh
import pandas as pd
import numpy as np

In [2]:
url='https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'

#Create a handle, page, to handle the contents of the website
page = requests.get(url)

#Store the contents of the website under doc
doc = lh.fromstring(page.content)

#Parse data that are stored between <tr>..</tr> of HTML
tr_elements = doc.xpath('//tr')

In [3]:
#For sanity check, ensure that all the rows have the same width. If not, we probably got something more than just the table.

#Check the length of the first 12 rows
[len(T) for T in tr_elements[:12]]

[3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3]

Means we have three column per row in the table

### Parse Table Header

In [4]:
tr_elements = doc.xpath('//tr')
#Create empty list
col=[]
i=0
#For each row, store each first element (header) and an empty list
for t in tr_elements[0]:
    i+=1
    name=t.text_content()
    print( '%d:"%s"'%(i,name))
    col.append((name,[]))

1:"Postcode"
2:"Borough"
3:"Neighbourhood
"


### Creating Pandas DataFrame
Each header is appended to a tuple along with an empty list

In [5]:
#Since out first row is the header, data is stored on the second row onwards
for j in range(1,len(tr_elements)):
    #T is our j'th row
    T=tr_elements[j]
    
    #If row is not of size 3, the //tr data is not from our table 
    if len(T)!=3:
        break
    
    #i is the index of our column
    i=0
    
    #Iterate through each element of the row
    for t in T.iterchildren():
        data=t.text_content() 
        #Check if row is empty
        if i>0:
        #Convert any numerical value to integers
            try:
                data=int(data)
            except:
                pass
        #Append the data to the empty list of the i'th column
        col[i][1].append(data)
        #Increment i for the next column
        i+=1

In [6]:
# Just to be sure, let’s check the length of each column. Ideally, they should all be the same.

[len(C) for (title,C) in col]

[287, 287, 287]

This shows that each of the 3 columns has exactly 287 rows



### Creating the pandas data fram

In [7]:
Dict={title:column for (title,column) in col}
df=pd.DataFrame(Dict)

In [8]:
df.head()

Postcode           Borough     Neighbourhood\n
0      M1A      Not assigned      Not assigned\n
1      M2A      Not assigned      Not assigned\n
2      M3A        North York         Parkwoods\n
3      M4A        North York  Victoria Village\n
4      M5A  Downtown Toronto      Harbourfront\n

### Data Cleaning and Preprocessing 

In [9]:
# Remove end of line \n from data
df = df.replace('\n',' ', regex=True)
df.head()

Postcode           Borough    Neighbourhood\n
0      M1A      Not assigned      Not assigned 
1      M2A      Not assigned      Not assigned 
2      M3A        North York         Parkwoods 
3      M4A        North York  Victoria Village 
4      M5A  Downtown Toronto      Harbourfront

In [10]:
# Remove Missing (not assigned) data
df.drop(df.index[df['Borough'] == 'Not assigned'], inplace = True)

# Reset the index and dropping the previous index
df = df.reset_index(drop=True)

df.head(10)

Postcode           Borough    Neighbourhood\n
0      M3A        North York         Parkwoods 
1      M4A        North York  Victoria Village 
2      M5A  Downtown Toronto      Harbourfront 
3      M6A        North York  Lawrence Heights 
4      M6A        North York    Lawrence Manor 
5      M7A      Queen's Park      Not assigned 
6      M9A      Queen's Park      Queen's Park 
7      M1B       Scarborough             Rouge 
8      M1B       Scarborough           Malvern 
9      M3B        North York   Don Mills North

In [11]:
df.rename(columns={'Neighbourhood\n':'Neighbourhood'}, inplace=True)
df.head()

Postcode           Borough      Neighbourhood
0      M3A        North York         Parkwoods 
1      M4A        North York  Victoria Village 
2      M5A  Downtown Toronto      Harbourfront 
3      M6A        North York  Lawrence Heights 
4      M6A        North York    Lawrence Manor

In [12]:
#Combining Neighbourhoods based on similar Postcode and Borough
df = df.groupby(['Postcode', 'Borough'])['Neighbourhood'].apply(','.join).reset_index()
df.columns = ['Postcode','Borough','Neighbourhood']
df.head()

Postcode      Borough                            Neighbourhood
0      M1B  Scarborough                          Rouge ,Malvern 
1      M1C  Scarborough  Highland Creek ,Rouge Hill ,Port Union 
2      M1E  Scarborough       Guildwood ,Morningside ,West Hill 
3      M1G  Scarborough                                  Woburn 
4      M1H  Scarborough                               Cedarbrae

In [13]:
df['Neighbourhood'] = df['Neighbourhood'].str.strip()
df.head()

Postcode      Borough                           Neighbourhood
0      M1B  Scarborough                          Rouge ,Malvern
1      M1C  Scarborough  Highland Creek ,Rouge Hill ,Port Union
2      M1E  Scarborough       Guildwood ,Morningside ,West Hill
3      M1G  Scarborough                                  Woburn
4      M1H  Scarborough                               Cedarbrae

In [14]:
#Assigning Borough values to the Neignbourhood where vlaue is "Not assigned"

df.loc[df['Neighbourhood'] == 'Not assigned', 'Neighbourhood'] = df['Borough']

In [15]:
# Verifying for Queen's Park changed 
df[df['Borough'] == 'Queen\'s Park']

Postcode       Borough Neighbourhood
85      M7A  Queen's Park  Queen's Park
93      M9A  Queen's Park  Queen's Park

In [16]:
df.shape

(103, 3)

Now that you have built a dataframe of the postal code of each neighborhood along with the borough name and neighborhood name, in order to utilize the Foursquare location data, we need to get the latitude and the longitude coordinates of each neighborhood.

In an older version of this course, we were leveraging the Google Maps Geocoding API to get the latitude and the longitude coordinates of each neighborhood. However, recently Google started charging for their API: http://geoawesomeness.com/developers-up-in-arms-over-google-maps-api-insane-price-hike/, so we will use the Geocoder Python package instead: https://geocoder.readthedocs.io/index.html.

The problem with this Package is you have to be persistent sometimes in order to get the geographical coordinates of a given postal code. So you can make a call to get the latitude and longitude coordinates of a given postal code and the result would be None, and then make the call again and you would get the coordinates. So, in order to make sure that you get the coordinates for all of our neighborhoods, you can run a while loop for each postal code. Taking postal code M5G as an example, your code would look something like this:


Given that this package can be very unreliable, in case you are not able to get the geographical coordinates of the neighborhoods using the Geocoder package, here is a link to a csv file that has the geographical coordinates of each postal code: http://cocl.us/Geospatial_data

Use the Geocoder package or the csv file to create the following dataframe:



## Using CSV file

In [17]:
url2 = "http://cocl.us/Geospatial_data"
df2 = pd.read_csv(url2)

df2.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [18]:
#checking df2 shape

df2.shape

(103, 3)

df and df2 have the same shape

### Merging both dataframes

In [19]:
# Rename Postal Code to Postcode
df2.rename(columns={'Postal Code':'Postcode'}, inplace=True)
df2.head()

Postcode   Latitude  Longitude
0      M1B  43.806686 -79.194353
1      M1C  43.784535 -79.160497
2      M1E  43.763573 -79.188711
3      M1G  43.770992 -79.216917
4      M1H  43.773136 -79.239476

In [21]:
# Merging 
df_merge = pd.merge(df, df2, on='Postcode')
df_merge.head()

Postcode      Borough                           Neighbourhood   Latitude  \
0      M1B  Scarborough                          Rouge ,Malvern  43.806686   
1      M1C  Scarborough  Highland Creek ,Rouge Hill ,Port Union  43.784535   
2      M1E  Scarborough       Guildwood ,Morningside ,West Hill  43.763573   
3      M1G  Scarborough                                  Woburn  43.770992   
4      M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476